### PR TITLE
Wait for amp-selector to build before submitting

### LINKED ISSUE #6927: amp-form should wait for amp-selector to finish loading before submission
This is a followup issue to keep track of `amp-selector` and `amp-form` submission.

Splitting these into different PRs. First PR will address waiting for `amp-form` through delegating XHR submissions through action service and the next PR will address waiting for `amp-selector`.

### LINKED ISSUE #7032: Implement a build promise that resolve when an element builds
As a follow up for #6951 we need a way to know when an element has been built.

This is because waiting for an extension to download still allows for a small chance of a race condition where we execute submission after the extension has installed but before elements have been built.

### PR DESCRIPTION
**Pull request by mkhatib**
Closes #7032 
Closes #6951
Closes #6927

### CODE REVIEW

File: extensions/amp-form/0.1/amp-form.js [hunk 1]
```diff
@@ -53,6 +53,15 @@ import {
 const TAG = 'amp-form';
 
 
+/**
+ * A list of external dependencies that can be included in forms.
+ * @type {Array<string>}
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: `!Array<string>`

File: extensions/amp-form/0.1/amp-form.js [hunk 2]
```diff
@@ -181,10 +188,24 @@ export class AmpForm {
...
+   * tag are loaded and built (e.g. amp-selector) or 500ms timeout - whichever is first.
+   * @return {!Promise}
+   * @private
+   */
+  whenDependenciesReady_() {
+    const depElements = this.form_.querySelectorAll(EXTERNAL_DEPS.join(','));
+    // Wait for an element to be built to make sure it is ready.
+    const depPromises = toArray(depElements).map(el => el.whenBuilt());
+    return Promise.race(
+        [Promise.all(depPromises), this.timer_.promise(500)]);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: 500 is not enough. Let's go with at least 2s

File: src/custom-element.js [hunk 3]
```diff
@@ -448,6 +448,18 @@ function createBaseCustomElementClass(win) {
...
       this.built_ = false;
 
+      /** @private {?function()} */
+      this.buildPromiseResolver_ = null;
+
+      /** @private {?function(!Error)} */
+      this.buildPromiseRejector_ = null;
+
+      /** @private {!Promise} */
+      this.buildPromise_ = new Promise((resolve, reject) => {
```
Comment (dvoytenko) [inline src/custom-element.js]: Can we make this lazy?

Comment (mkhatib) [inline src/custom-element.js]: You mean creating the promise? Yes sure.

Comment (mkhatib) [inline src/custom-element.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 2 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

File: extensions/amp-form/0.1/amp-form.js (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

**Comment by mkhatib** [comment]
@dvoytenko PTAL 👀

**Comment by jridgewell** [comment]
> Closes #6951
Closes #6951 

Are there supposed to be two separate issues there?

File: extensions/amp-form/0.1/amp-form.js [hunk 4]
```diff
@@ -181,10 +188,24 @@ export class AmpForm {
...
+   * Returns a promise that will be resolved when all dependencies used inside the form
+   * tag are loaded and built (e.g. amp-selector) or 500ms timeout - whichever is first.
+   * @return {!Promise}
+   * @private
+   */
+  whenDependenciesReady_() {
+    const depElements = this.form_.querySelectorAll(EXTERNAL_DEPS.join(','));
+    // Wait for an element to be built to make sure it is ready.
+    const depPromises = toArray(depElements).map(el => el.whenBuilt());
+    return Promise.race(
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Can we cache this, so we don't have to re-query every time?

File: src/custom-element.js [hunk 5]
```diff
@@ -669,6 +678,33 @@ function createBaseCustomElementClass(win) {
...
+     * Initialize the build promise if hasn't been already done.
+     * @return {!Promise}
+     * @private
+     */
+    initBuildPromise_() {
+      if (this.buildPromise_) {
+        return this.buildPromise_;
+      }
+      return this.buildPromise_ = new Promise((resolve, reject) => {
+        this.buildPromiseResolver_ = resolve;
```
Comment (jridgewell) [inline src/custom-element.js]: There's no reason to hold onto these references is we're already built.

File: src/custom-element.js [hunk 6]
```diff
@@ -669,6 +678,33 @@ function createBaseCustomElementClass(win) {
...
+        if (this.built_) {
```
Comment (jridgewell) [inline src/custom-element.js]: We need to know whether the build failed as well.

File: src/custom-element.js (hunk 5 above)
Comment (mkhatib) [inline src/custom-element.js]: This logic is now changed - @dvoytenko implemented a signal system in #7205

File: src/custom-element.js (hunk 6 above)
Comment (mkhatib) [inline src/custom-element.js]: ditto

**Comment by jridgewell** [comment]
Let's rename the PR.

File: extensions/amp-form/0.1/amp-form.js (hunk 4 above)
Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Done

**Comment by mkhatib** [comment]
@dvoytenko & @jridgewell PTAL 👀

File: extensions/amp-form/0.1/amp-form.js [hunk 7]
```diff
@@ -181,8 +191,25 @@ export class AmpForm {
...
+   */
+  whenDependenciesReady_() {
+    if (this.dependenciesPromise_) {
+      return this.dependenciesPromise_;
     }
+    const depElements = this.form_.querySelectorAll(EXTERNAL_DEPS.join(','));
+    // Wait for an element to be built to make sure it is ready.
+    const depPromises = toArray(depElements).map(el => el.whenBuilt());
+    return this.dependenciesPromise_ = Promise.race(
+        [Promise.all(depPromises), this.timer_.promise(15000)]);
```
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: 15 seconds?

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Woops good catch I was testing this and forgot to reset it to 2000

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Fixed

File: extensions/amp-form/0.1/amp-form.js [hunk 8]
```diff
@@ -181,8 +191,25 @@ export class AmpForm {
...
   actionHandler_(invocation) {
     if (invocation.method == 'submit') {
-      this.handleSubmitAction_();
+      this.whenDependenciesReady_().then(this.handleSubmitAction_.bind(this));
+    }
+  }
+
+  /**
+   * Returns a promise that will be resolved when all dependencies used inside the form
+   * tag are loaded and built (e.g. amp-selector) or 500ms timeout - whichever is first.
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Update jsdoc.

File: extensions/amp-form/0.1/amp-form.js [hunk 9]
```diff
@@ -104,6 +113,9 @@ export class AmpForm {
     /** @const @private {!../../../src/service/url-replacements-impl.UrlReplacements} */
     this.urlReplacement_ = urlReplacementsForDoc(element);
 
+    /** @private {?Promise} */
+    this.dependenciesPromise_ = null;
```
Comment (camelburrito) [inline extensions/amp-form/0.1/amp-form.js]: nit(optional): i would call this - haveDepsLoadedPromise_

Comment (mkhatib) [inline extensions/amp-form/0.1/amp-form.js]: Sticking to old name, the method name is more important as it describes the wait better.

**Pull request merged**